File: 2.两数相加.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] 两数相加
#

# @lc code=start
# Definition for singly-linked list.
class ListNode:
     def __init__(self, val=0, next=None):
         self.val = val
         self.next = next
from multiprocessing import dummy


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy=ListNode()
        cur=dummy
        carry=0
        while l1 or l2 or carry:
            v1=l1.val if l1 else 0
            v2=l2.val if l2 else 0
            val=v1+v1+carry

            carry=val//10
            cur.next=ListNode(val%10)

            cur=cur.next
            l1=l1.next if l1 else None
            l2=l2.next if l2 else None
        
        return dummy.next

# @lc code=end

File: 2.两数相加.java
/*
 * @lc app=leetcode.cn id=2 lang=java
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode cur = dummy;
        int carry = 0;
        while(l1!=null&&l2!=null){
            int val = (l1.val+l2.val+carry)%10;
            carry = (l1.val+l2.val+carry)/10;
            cur.next = new ListNode(val);
            cur = cur.next;
            l1=l1.next;
            l2=l2.next;
        }

        while(l1!=null){
            int val = (l1.val+carry)%10;
            carry = (l1.val+carry)/10;
            cur.next = new ListNode(val);
            cur = cur.next;
            l1=l1.next;
        }

         while(l2!=null){
            int val = (l2.val+carry)%10;
            carry = (l2.val+carry)/10;
            cur.next = new ListNode(val);
            cur = cur.next;
            l2=l2.next;
        }

        if(carry == 1){
            cur.next = new ListNode(1);
        }

        return dummy.next;

    }
}
// @lc code=end

